File: drills/ts/solutions/nc123_alien_dictionary__dfs_postorder.ts
export function alienOrder(words: string[]): string {
  const letters = new Set<string>();
  for (const word of words) for (const letter of word) letters.add(letter);

  const after = new Map<string, string[]>();
  for (const letter of letters) after.set(letter, []);

  for (let i = 0; i + 1 < words.length; i++) {
    const first = words[i];
    const second = words[i + 1];
    const shared = Math.min(first.length, second.length);
    let split = -1;
    for (let j = 0; j < shared; j++) {
      if (first[j] !== second[j]) {
        split = j;
        break;
      }
    }
    if (split === -1) {
      if (first.length > second.length) return "";
      continue;
    }
    after.get(first[split])!.push(second[split]);
  }

  // Depth-first, recording a letter only once everything that must follow it
  // has been recorded -- so the record comes out backwards and is reversed at
  // the end. The in-progress set is the cycle check: a letter met again on the
  // current path contradicts itself.
  const onPath = new Set<string>();
  const done = new Set<string>();
  const order: string[] = [];

  const visit = (letter: string): boolean => {
    if (onPath.has(letter)) return false;
    if (done.has(letter)) return true;
    onPath.add(letter);
    for (const following of after.get(letter)!) {
      if (!visit(following)) return false;
    }
    onPath.delete(letter);
    done.add(letter);
    order.push(letter);
    return true;
  };

  for (const letter of letters) {
    if (!visit(letter)) return "";
  }

  return order.reverse().join("");
}

File: drills/go/solutions/nc123_alien_dictionary__dfs_postorder.go
package main

func alienOrder(words []string) string {
	next := map[byte]map[byte]bool{}
	for _, w := range words {
		for i := 0; i < len(w); i++ {
			if next[w[i]] == nil {
				next[w[i]] = map[byte]bool{}
			}
		}
	}
	for i := 0; i+1 < len(words); i++ {
		a, b := words[i], words[i+1]
		if len(a) > len(b) && a[:len(b)] == b {
			return ""
		}
		for j := 0; j < len(a) && j < len(b); j++ {
			if a[j] != b[j] {
				next[a[j]][b[j]] = true
				break
			}
		}
	}
	// Depth-first post-order with path colouring: a letter is emitted after
	// everything that must follow it, so the reversed order is the answer.
	const unvisited, onPath, done = 0, 1, 2
	colour := map[byte]int{}
	postorder := []byte{}
	var visit func(c byte) bool
	visit = func(c byte) bool {
		switch colour[c] {
		case onPath:
			return false
		case done:
			return true
		}
		colour[c] = onPath
		for d := byte('a'); d <= 'z'; d++ {
			if next[c][d] && !visit(d) {
				return false
			}
		}
		colour[c] = done
		postorder = append(postorder, c)
		return true
	}
	for c := byte('a'); c <= 'z'; c++ {
		if _, present := next[c]; present && !visit(c) {
			return ""
		}
	}
	for i, j := 0, len(postorder)-1; i < j; i, j = i+1, j-1 {
		postorder[i], postorder[j] = postorder[j], postorder[i]
	}
	return string(postorder)
}
